napari/_vispy/visuals/volume.py

```diff
@@ -113,6 +113,34 @@
 }
 """
 
+ATTENUATED_MIP_SNIPPETS = dict(
+    before_loop="""
+        float maxval = u_mip_cutoff; // The maximum encountered value
+        float scaled_sumval = 0.0; // The sum of the encountered values
+        float scaled = 0.0; // The scaled value
+        int maxi = -1;  // Where the maximum value was encountered
+        vec3 max_loc_tex = vec3(0.0);  // Location where the maximum value was encountered
+        """,
+    in_loop="""
+        scaled_sumval += (val - clim.x) / (clim.y - clim.x);
```

Comment:
This is the relevant change to the shader, otherwise it's just copied in from vispy.

Reply:
Great! We should probably upstream this to vispy as well, so we can drop the vendoring here.

Reply:
Now that I think about it, does this mean we can/should fix the isosurface shader in the same way, instead of doing stuff in python?

Reply:
Yeah taking a look at the `in_loop` portion of the isosurface shader, it seems like it's assuming something about the scale of the data with this line:
```
if (val > u_threshold-0.2) {
    ...
```

I'll try some similar changes there - it would simplify the code in napari and unify the iso and attenuated_mip implementations.

Reply:
Actually it seems this might not work. Only the contrast limits–not the range–are passed to the shader, so I can't get the same scaling in the shader without more changes to vispy (and I'm not sure making those changes would make sense).

Reply:
I see... Yeah maybe this is overcomplicating things unnecessarily :P